File: houserobberii.js
//Objective is the same as 'House Robber', except this time the first and
//last house in the list are also adjacent.

let nums = [2,3,2]


//O(n) solution that uses two dp arrays that start from the beginning and one
//after the beginning, then compares to the two arrays

if (nums.length < 2) {
    return nums[0] || 0
}

let start = [nums[0]]
let after = [0, nums[1]]

for (let i = 1; i < nums.length - 1; i++) {
    start[i] = Math.max(start[i - 1], (start[i - 2] || 0) + nums[i])
}

for (let i = 2; i < nums.length; i++) {
    after[i] = Math.max(after[i - 1], after[i - 2] + nums[i])
}

return Math.max(start.pop(), after.pop())

File: houserobberii.java
//Java Solution

class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        
        int[] start = new int[nums.length];
        start[0] = nums[0];
        
        int[] end = new int[nums.length];
        end[0] = 0;
        end[1] = nums[1];
        
        for (int i = 1; i < nums.length - 1; i++) {
            start[i] = (i >= 2) ? Math.max(start[i - 1], start[i - 2] + nums[i]) : Math.max(start[i - 1], nums[i]);
        }
        
        for (int i = 2; i < nums.length; i++) {
            end[i] = Math.max(end[i - 1], end[i - 2] + nums[i]);
        }
        
        return Math.max(start[start.length - 2], end[end.length - 1]);
    }
}